### PR TITLE
refactor: instruments build and framework cleanup

### PR DESCRIPTION
**Pull request by Benjozork**
## Summary of Changes

* [x] Use parallelized build for instruments
* [x] Enable support for TypeScript path mappings (depends on https://github.com/simonhaenisch/rollup-plugin-typescript-paths/pull/9)
* [x] Enable use of `rollup.config.js` regardless of cwd
* [x] Separate simulator and browser builds
* [ ] Add `devServer`
  * [x] Base 
  * [ ] `SimVar` shim
* [x] Adapt A32NX build scripts

Discord username (if different from GitHub): someperson#4953

## Testing instructions

* Ensure all instruments and ECAM pages work

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by HPringles** [comment]
**Category**: QA Trial
**Name**: HPringles#7078
**Date of testing**: 23/05/2020
**Version of the sim**: 10.
**PR Tested**: #4750
**Tier of testing**: 1
### Changes to observe:
Build Works & therefore so do displays

### Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Started at gate Cold&Dark
2. Turned batteries/gnd pwr on

5.

**Issues (If Any):** React doesn't load on the PFD
### Steps to achieve the issue
Look at the PFD

image

image

**Overall Rating:** Not quite there
**Conclusions of the testing**: Needs a fix :(

**Comment by ItzCryptiic** [comment]
Category: QA Trial
Name: Cryptiic#8175
Date of testing: 18/05/2021
Version of the sim: 1.15.10.0
PR Tested: #4750

Tier of testing: 1

Changes to observe: Ensure all instruments and ECAM pages work

Steps you've made to test the Pull Request:

  1. Spawned Cold & Dark at EGCC
  2. Started aircraft up (Before Start-up Checklist)
  3. Checked every Display (PFD, ND's, every ECAM page, flyPad, FCU, Communication Panel und Squawk Box)

Flight notes (When done a full flight): N/A
  
Media: 
Screenshot (170)
Screenshot (171)
Screenshot (172)
Screenshot (173)
Screenshot (174)

Issues (If Any):
On the PFD, flyPad, Communication Panel and Squawk Box I got the same thing displayed a line of text saying: If you're seeing this, React didn't load. However, the right ND, MCDU, FCU and every ECAM Page worked normally.

Overall Rating: Needs Improvement
Conclusions of the testing: PFD, flyPad, Communication Panel and Squawk Box weren't working. ND, MCDU, FCU and ECAM worked normally.

**Comment by HPringles** [comment]
**Category**: QA Trial
**Name**: HPringles#7078
**Date of testing**: 23/05/2020
**Version of the sim**: 10.
**PR Tested**: #4750
**Tier of testing**: 1
### Changes to observe:
Build Works & therefore so do displays

### Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Started at gate Cold&Dark
2. Turned batteries/gnd pwr on
3. Check Pages
5. They Work

**Issues (If Any):** None - all resolved
### Steps to achieve the issue






**Overall Rating:** Good
**Conclusions of the testing**: LGTM

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 23/5/21
Version of the sim: 1.15.10.0
PR Tested: #4750 
Tier of testing: 1/2/3
Changes to observe:
black displays

Testing technique: Loaded in and tested
Steps you've made to test the Pull Request:
1. Loaded C&D
2. looked at all displays

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.

Issues (If Any): none
Steps to achieve the issue

Overall Rating: Good
Conclusions of the testing: LGTM

**Comment by scoopscode** [comment]
Category: QA Tester
Name: Scoopy#1399
Date of testing: 23/5/21
Version of the sim: 1.15.10.0
PR Tested: #4750
Tier of testing: 1/2/3
Changes to observe:
If blacked out displays are functioning again

Testing technique: Loaded in and tested
Steps you've made to test the Pull Request:

Loaded C&D
looked at all displays
Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.
Screenshot (46)
Screenshot (47)
Screenshot (45)


Issues (If Any): none
Steps to achieve the issue

Overall Rating: Good
Conclusions of the testing: Working as expected!

**Comment by ItzCryptiic** [comment]
Category: QA Trial
Name: Cryptiic#8175
Date of testing: 23/05/2021
Version of the sim: 1.15.10.0
PR Tested: #4750

Tier of testing: 1

Changes to observe: Ensure all instruments and ECAM pages work - Test 2

Steps you've made to test the Pull Request:

1. Spawned Cold & Dark at EGCC 
2. Started aircraft up (Before Start-up Checklist)
3. Checked every Display (PFD, ND's, every ECAM page, flyPad, FCU, Communication Panel und Squawk Box)

Flight notes (When done a full flight): N/A

Media: N/A

Issues (If Any):
No Issues regarding Ensure all instruments and ECAM pages work found.

Overall Rating: Good
Conclusions of the testing: LGTM

File: package.json [hunk 1]
```diff
@@ -5,10 +5,11 @@
     "lint": "eslint --cache **/*.{js,mjs,jsx,ts,tsx}",
     "lint-fix": "npm run lint -- --fix",
     "prettier": "prettier --write **/*.json **/*.yml src/instruments/**/*.css",
-    "build:instruments": "rollup -c src/instruments/rollup.config.js",
+    "build:instruments": "cd src/instruments && ./build.sh",
     "serve:efb": "cd src/instruments/src/EFB/ && rollup -wc ./rollup.development.config.js",
     "build:fbw": "cd src/fbw && ./build.sh",
-    "build:fadec": "cd src/fadec && ./build.sh"
+    "build:fadec": "cd src/fadec && ./build.sh",
+    "build:fmgc": "rollup -c src/fmgc/rollup.config.js"
```
[the remark this answers is not in the document]
Comment (Benjozork) [reply package.json:12]: yep

File: package.json [hunk 2]
```diff
@@ -41,6 +43,7 @@
     "awesome-typescript-loader": "^5.2.1",
     "css-loader": "^5.0.1",
     "eslint": "7.21.0",
+    "fancy-rollup": "^0.5.5",
```
[the remark this answers is not in the document]
Comment (Benjozork) [reply package.json:46]: no

File: src/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    "baseUrl": ".",
+    "allowJs": true,
+    "paths": {
+      "@shared/*": ["./shared/src/*"],
+      "@instruments/common/*": ["./instruments/src/Common/*"],
+      "@fmgc/*": ["./fmgc/src/*"],
+      "@typings/*": ["../typings/*"]
+    }
+  }
+}
```
Comment (theomessin) [inline src/tsconfig.json:12]: Trailing new line.

File: .eslintignore [hunk 4]
```diff
@@ -3,4 +3,5 @@ flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/math.js
 build.*.js
 buildSrc/*
 igniter.config.mjs
-src/instruments/ecam-page-template/template.js
\ No newline at end of file
+src/instruments/ecam-page-template/template.js
+src/instruments/devServer/bundles/*
```
Comment (theomessin) [inline .eslintignore:7]: Trailing new line.

File: package.json [hunk 5]
```diff
@@ -5,7 +5,7 @@
     "lint": "eslint --cache **/*.{js,mjs,jsx,ts,tsx}",
     "lint-fix": "npm run lint -- --fix",
     "prettier": "prettier --write **/*.json **/*.yml src/instruments/**/*.css",
-    "build:instruments": "rollup -c src/instruments/rollup.config.js",
+    "build:instruments": "cd src/instruments && ./build.sh",
     "serve:efb": "cd src/instruments/src/EFB/ && rollup -wc ./rollup.development.config.js",
     "build:fbw": "cd src/fbw && ./build.sh",
     "build:fadec": "cd src/fadec && ./build.sh"
```
Comment (theomessin) [inline package.json:11]: Why are these needed? This can be done by passing a regex to igniter instead...

File: src/instruments/build.sh [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+rollup -c src/instruments/buildSrc/simulatorBuild.mjs
```
Comment (theomessin) [inline src/instruments/build.sh:3]: Why do we need a separate `.sh` script for this?

File: src/instruments/buildSrc/igniter/RollupTask.mjs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+import { exec } from 'child_process';
+
+export class RollupTask extends GenericTask {
+    constructor(
+        key,
+        targetToBuild,
+        hashFolders,
+    ) {
+        super(key, async () => {
+            await promisify(exec)(`node src/instruments/buildSrc/igniter/RollupTaskWorker.mjs ${targetToBuild}`);
```
Comment (theomessin) [inline src/instruments/buildSrc/igniter/RollupTask.mjs:12]: Why not just have the task be the code in `RollupTaskWorker` as opposed to running the code using node? (i.e. merge the two files)

File: src/instruments/devServer/index.html [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+            </div>
+            <div id="Display">
+
+            </div>
+        </div>
+
+        <script defer src="js/loader.js"></script>
+    </body>
+</html>
```
Comment (theomessin) [inline src/instruments/devServer/index.html:22]: Trailing new line.

File: src/instruments/devServer/styles.css [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+.SidebarItem {
+  height: 3em;
+
+  display: flex;
+  flex-direction: row;
+  justify-content: flex-start;
+  align-items: center;
+
+  padding-left: 1em;
+}
```
Comment (theomessin) [inline src/instruments/devServer/styles.css:26]: Trailing new line.

**Review by theomessin — COMMENTED** [review]
I'm a bit confused with the way the ECAM pages work so I haven't reviewed that properly.
Other than that, I really like the new paths and use of Igniter 😄 
Have some minor comments.

FFR, might be easier to break into smaller PRs.

File: src/instruments/buildSrc/igniter/RollupTask.mjs (hunk 7 above)
Comment (Benjozork) [reply src/instruments/buildSrc/igniter/RollupTask.mjs:12]: It wouldn't run asynchronously then would it ?

File: src/instruments/build.sh (hunk 6 above)
Comment (Benjozork) [reply src/instruments/build.sh:3]: Forgot to remove it.

File: package.json (hunk 5 above)
Comment (Benjozork) [reply package.json:11]: Forgot to remove them

File: src/instruments/buildSrc/igniter/RollupTask.mjs (hunk 7 above)
Comment (theomessin) [reply src/instruments/buildSrc/igniter/RollupTask.mjs:12]: Good catch, yes. You'd have to use [worker threads](https://nodejs.org/api/worker_threads.html). Up to you if you wanna keep exec. I'm happy either way

**Comment by Benjozork** [comment]
Fixed.

**Review by lukecologne — APPROVED** [review]

**Pull request merged**